### PR TITLE
fix(Core/Script): 'Battle at Valhalas' quests done automatically

### LINKED ISSUE #6753: Battle at Valhalas quests,it's done automatically [$15]
### Current Behaviour

`Battle at Valhalas` series of quests, get to automatically complete

### Expected Blizzlike Behaviour

get `Battle at Valhalas` series of quests,defeat the summoned NPC before completing the quest.

### Source

_No response_

### Steps to reproduce the problem

1.  .quest add 13213
2.  find `Geirrvif` (.go creature id 31135)and then complete quest 13213
3. get quest `Battle at Valhalas: Fallen Heroes `(id 13214), don't  need kill any npc.
4. find `Gjonner the Merciless` (.go creature id 31154),then complete quest `Battle at Valhalas: Fallen Heroes `(id 13214)
5. find `Geirrvif` (.go creature id 31135)and get quest `Battle at Valhalas: Khit'rix the Dark Master`(id 13215),don't  need kill any npc.
6.  find `Gjonner the Merciless` (.go creature id 31154),then complete quest `Battle at Valhalas: Khit'rix the Dark Master`(id 13215)
7. find `Geirrvif` (.go creature id 31135)and get quest `Battle at Valhalas: The Return of Sigrid Iceborn`(id 13216),don't  need kill any npc.
8.  find `Gjonner the Merciless` (.go creature id 31154),then complete quest `Battle at Valhalas: The Return of Sigrid Iceborn`(id 13216)
9.  find `Geirrvif` (.go creature id 31135)and get quest `Battle at Valhalas: Carnage!`(id 13217),don't  need kill any npc.
10. find `Gjonner the Merciless` (.go creature id 31154),then complete quest `Battle at Valhalas: Carnage!`(id 13217)
11. find `Geirrvif` (.go creature id 31135)and get quest `Battle at Valhalas: Thane Deathblow`(id 13218),don't  need kill any npc.
12.  find `Gjonner the Merciless` (.go creature id 31154),then complete quest `Battle at Valhalas: Thane Deathblow`(id 13218)
13. find `Geirrvif` (.go creature id 31135)and get quest `Battle at Valhalas: Final Challenge`(id 13219),don't  need kill any npc.
14. find `Gjonner the Merciless` (.go creature id 31154),then complete quest `Battle at Valhalas: Final Challenge`(id 13219)

### Extra Notes

_No response_

### AC rev. hash/commit

47b09f27671d9d04e86cfc996085207c8b5ac177

### Operating system

ubuntu20.04

### Custom changes or Modules

_No response_

<bountysource-plugin>

---
There is a **[$15 open bounty](https://www.bountysource.com/issues/99776021-battle-at-valhalas-quests-it-s-done-automatically?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github)** on this issue. Add to the bounty at [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by Zenslash**
## Changes Proposed:
-  Previously, quests were done automatically because QuestGiver's data was not reset after the previous attempt. Now it fixed
-  Also, the cause of the problem was ```TEMPSUMMON_TIMED_DESPAWN_OUT_OF_COMBAT``` flag. If the player died, then all npc despawned. After that, QuestGiver checked whether all the npc were alive, and since they despawned after player death, the quest was considered completed. Now the npc will not despawn until the player kills them or cancels the quest

## Issues Addressed:
- Closes #6753 

## Tests Performed:
- Tested on local server

## How to Test the Changes:
1. ```.quest add 13213```
2. find ```Geirrvif``` (.go creature id 31135)and then complete quest 13213
3. get quest ```Battle at Valhalas: Fallen Heroes``` (id 13214)
4. find ```Gjonner the Merciless``` (.go creature id 31154),then complete quest ```Battle at Valhalas: Fallen Heroes``` (id 13214)
5. find ```Geirrvif``` (.go creature id 31135)and get quest ```Battle at Valhalas: Khit'rix the Dark Master``` (id 13215)
6. find ```Gjonner the Merciless``` (.go creature id 31154),then complete quest ```Battle at Valhalas: Khit'rix the Dark Master``` (id 13215)
7. find ```Geirrvif``` (.go creature id 31135)and get quest ```Battle at Valhalas: The Return of Sigrid Iceborn``` (id 13216)
8. find ```Gjonner the Merciless``` (.go creature id 31154),then complete quest ```Battle at Valhalas: The Return of Sigrid Iceborn``` (id 13216)
9. find ```Geirrvif``` (.go creature id 31135)and get quest ```Battle at Valhalas: Carnage!``` (id 13217)
10. find ```Gjonner the Merciless``` (.go creature id 31154),then complete quest ```Battle at Valhalas: Carnage!``` (id 13217)
11. find ```Geirrvif``` (.go creature id 31135)and get quest ```Battle at Valhalas: Thane Deathblow``` (id 13218)
12. find ```Gjonner the Merciless``` (.go creature id 31154),then complete quest ```Battle at Valhalas: Thane Deathblow``` (id 13218)
13. find ```Geirrvif``` (.go creature id 31135)and get quest ```Battle at Valhalas: Final Challenge``` (id 13219)
14. find ```Gjonner the Merciless``` (.go creature id 31154),then complete quest ```Battle at Valhalas: Final Challenge``` (id 13219)

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

File: src/server/game/Entities/Player/Player.cpp [hunk 1]
```diff
@@ -11485,6 +11485,65 @@ bool Player::HasQuestForGO(int32 GOId) const
...
+                    obj->BuildValuesUpdateBlockForPlayer(&udata, this);
+                    break;
+                }
+            }
+        }
+    }
+    udata.BuildPacket(&packet);
+    GetSession()->SendPacket(&packet);
+}
+
```
Comment (FrancescoBorzi) [inline src/server/game/Entities/Player/Player.cpp:11546]: I guess this should be part of `PlayerQuest.cpp` instead

Comment (Zenslash) [reply src/server/game/Entities/Player/Player.cpp]: @FrancescoBorzi I removed multiple definition for ```UpdateForQuestWorldObjects()``` and moved ```HasQuest()``` to ```PlayerQuest.cpp```

File: src/server/game/Entities/Player/PlayerUpdates.cpp [hunk 2]
```diff
@@ -1821,32 +1820,23 @@ void Player::UpdateForQuestWorldObjects()
...
-                        obj->GetEntry(), _itr->second.spellId);
-                bool buildUpdateBlock = false;
-                for (ConditionList::const_iterator jtr = conds.begin();
-                     jtr != conds.end() && !buildUpdateBlock; ++jtr)
-                    if ((*jtr)->ConditionType == CONDITION_QUESTREWARDED ||
-                        (*jtr)->ConditionType == CONDITION_QUESTTAKEN)
+                //! This code doesn't look right, but it was logically converted to condition system to do the exact
+                //! same thing it did before. It definitely needs to be overlooked for intended functionality.
+                ConditionList conds            = sConditionMgr->GetConditionsForSpellClickEvent(obj->GetEntry(), _itr->second.spellId);
+                bool          buildUpdateBlock = false;
```
Comment (IntelligentQuantum) [inline src/server/game/Entities/Player/PlayerUpdates.cpp:1837]:
```suggestion
                ConditionList conds = sConditionMgr->GetConditionsForSpellClickEvent(obj->GetEntry(), _itr->second.spellId);
                bool buildUpdateBlock = false;
```

File: src/server/game/Entities/Player/PlayerQuest.cpp [hunk 3]
```diff
@@ -84,6 +84,18 @@ void Player::PrepareQuestMenu(ObjectGuid guid)
     }
 }
 
+bool Player::HasQuest(uint32 questId) const
+{
+    for (uint8 i = 0; i < MAX_QUEST_LOG_SIZE; ++i)
+    {
+        uint32 questid = GetQuestSlotQuestId(i);
+        if (questid == questId)
+            return true;
```
Comment (IntelligentQuantum) [inline src/server/game/Entities/Player/PlayerQuest.cpp:93]:
```suggestion
        if (questid == questId)
        {
            return true;
        }
```

**Review by IntelligentQuantum — CHANGES_REQUESTED** [review]

**Comment by IntelligentQuantum** [comment]
Tested, it works fine for me.

File: src/server/scripts/Northrend/zone_icecrown.cpp [hunk 4]
```diff
@@ -180,7 +185,7 @@ class npc_battle_at_valhalas : public CreatureScript
             {
                 uint32 quest = currentQuest;
                 if (Player* player = ObjectAccessor::GetPlayer(*me, playerGUID))
-                {
+                {
```
Comment (IntelligentQuantum) [inline src/server/scripts/Northrend/zone_icecrown.cpp:188]:
```suggestion
                {
```

**Review by IntelligentQuantum — CHANGES_REQUESTED** [review]

**Review by IntelligentQuantum — APPROVED** [review]

**Pull request merged**